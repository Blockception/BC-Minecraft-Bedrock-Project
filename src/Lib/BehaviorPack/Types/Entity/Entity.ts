import { MolangSet } from "../../../Molang/MolangSet";
import { DefinedUsing } from "../../../Types/DefinedUsing";
import { Identifiable } from "../../../Types/Identifiable";
import { Using } from "../../../Types/Used";

export interface Entity extends Identifiable {
  molang: MolangSet;

  groups: string[];
  events: string[];
  animations: DefinedUsing<string>;
}
