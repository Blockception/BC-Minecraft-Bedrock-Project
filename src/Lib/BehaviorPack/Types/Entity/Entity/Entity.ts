import { MolangSet } from "../../../../Molang/MolangSet";
import { DefinedUsing } from "../../../../Types/DefinedUsing";
import { Identifiable } from "../../../../Types/Identifiable";

export interface Entity extends Identifiable {
  molang: MolangSet;

  groups: string[];
  events: string[];
  animations: DefinedUsing<string>;
}
