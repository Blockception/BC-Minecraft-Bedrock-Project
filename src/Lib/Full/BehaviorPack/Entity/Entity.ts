import { ScriptContainer } from "../../Types/Script";

export interface EntityEventGroup {
  component_groups?: string[];
}

export interface EntityEvent {
  add?: EntityEventGroup;
  remove?: EntityEventGroup;
  sequence?: EntityEvent[];
  randomize?: (EntityEvent & { weight: number })[];
}

export interface EntityComponentContainer {
  [key: string]: any;
}

export interface EntityDescription extends ScriptContainer {
  identifier: string;
  is_spawnable?: true;
  is_summonable?: true;
}

export interface Entity {
  format_version: string;
  "minecraf:entity": {
    description: EntityDescription;
    component_groups: {
      [group: string]: EntityComponentContainer;
    };
    components: EntityComponentContainer;
    events: {};
  };
}
