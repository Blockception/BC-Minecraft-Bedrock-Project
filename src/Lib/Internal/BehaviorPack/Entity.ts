import { Map } from "../../Types/Map";
import { ScriptContainer } from "../Types/Script";

/**The interface that deals with entity events that add or remove component groups*/
export interface EntityEventGroup {
  /**The group of entity events*/
  component_groups?: string[];
}

/** */
export interface EntityEvent {
  /** */
  add?: EntityEventGroup;
  /** */
  remove?: EntityEventGroup;
  /** */
  sequence?: EntityEvent[];
  /** */
  randomize?: (EntityEvent & { weight: number })[];
}

/** */
export interface EntityComponentContainer {
  /** */
  [key: string]: any;
}

/** */
export interface EntityDescription extends ScriptContainer {
  /** */
  identifier: string;
  /** */
  is_spawnable?: true;
  /** */
  is_summonable?: true;
}

/** */
export interface Entity {
  /** */
  format_version: string;
  /** */
  "minecraf:entity": {
    /** */
    description: EntityDescription;
    /** */
    component_groups?: Map<EntityComponentContainer>;
    /** */
    components: EntityComponentContainer;
    /** */
    events?: Map<EntityEvent>;
  };
}

/** */
export namespace Entity {
  /**
   *
   * @param value
   */
  export function is(value: any): value is Entity {
    if (value && typeof value.format_version === "string" && typeof value["minecraf:entity"] === "object") {
      const b = value["minecraf:entity"];

      if (typeof b.description === "object" && typeof b.description.identifier === "string" && typeof b.components === "object") {
        return true;
      }
    }

    return false;
  }
}
