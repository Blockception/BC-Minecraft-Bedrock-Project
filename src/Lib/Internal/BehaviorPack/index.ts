/*	Auto generated	*/

export * from "./Animation";
export * from "./AnimationController";
export * from "./Block";
export * from "./Entity";
export * from "./Item";