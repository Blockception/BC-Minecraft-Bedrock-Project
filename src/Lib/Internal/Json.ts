import { jsonc } from "jsonc";
import { TextDocument } from "../Types/TextDocument/TextDocument";

/**
 *
 */
export namespace Json {
  /**
   *
   * @param doc
   * @returns
   */
  export function To<T>(doc: TextDocument | string): T | undefined {
    let out: T | undefined = undefined;

    const content = typeof doc === "object" ? doc.getText() : doc;
    if (content !== "") out = jsonc.parse(content);

    return out;
  }
}
