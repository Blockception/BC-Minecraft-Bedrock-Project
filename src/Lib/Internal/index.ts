/*	Auto generated	*/
export * as BehaviorPack from "./BehaviorPack/index";
export * as Comment from "./Comment/index";
export * as General from "./General/index";
export * as ResourcePack from "./ResourcePack/index";
export * as Types from "./Types/index";
export * from "./Json";