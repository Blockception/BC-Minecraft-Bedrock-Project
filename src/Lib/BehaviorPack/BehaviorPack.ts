import { MCProject } from "bc-minecraft-project";
import { Container } from "../Types/Container";
import { DataSet } from "../Types/DataSet";
import { Edu, Types, Vanilla } from "bc-minecraft-bedrock-vanilla-data";
import { Entity } from "./Types/Entity/Entity";

export class BehaviorPack implements Container {
  readonly Context: MCProject;

  public entities: DataSet<Entity, Types.BehaviorPack.Entity>;

  constructor(Context: MCProject | string) {
    this.Context = typeof Context === "object" ? Context : MCProject.loadSync(Context);

    this.entities = DataSet.createID(Vanilla.BehaviorPack.Entities, Edu.BehaviorPack.Entities, this);
  }
}
