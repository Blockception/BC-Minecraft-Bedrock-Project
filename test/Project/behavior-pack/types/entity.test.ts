import { Json } from "../../../../src/Lib/Internal/Json";
import { TextDocument } from "../../../../src/Lib/Types/TextDocument";
import * as Internal from "../../../../src/Lib/Internal/BehaviorPack/Entity";
import { Entity } from "../../../../src/Lib/Project/BehaviorPack";

const EntityJson: string = `{
  "format_version": "1.16.0",
  "minecraft:entity": {
    "description": {
      "identifier": "blockception:sheep",
      "is_spawnable": true,
      "is_summonable": true,
      "animations": {
        "controller": "controller.animation.chicken",
        "something": "controller.i.dont.exist"
      },
      "scripts": {
        "animate": ["controller"]
      }
    },
    "component_groups": {
      "self:group1": {
        "minecraft:type_family": {
          "family": ["animal"]
        }
      }
    },
    "components": {
      "minecraft:health": { "value": 10, "max": 10 },
      "minecraft:type_family": {
        "family": ["mob"]
      }
    },
    "events": {
      "self:to": {}
    }
  }
}`;

const EntityDoc: TextDocument = {
  uri: "C:\\temp.json",
  getText: () => EntityJson,
};

describe("Entity", () => {
  describe("Data", () => {
    const imp = Json.To<Internal.Entity>(EntityJson);

    it("Not Undefined", () => {
      expect(imp).toBeDefined();
    });

    it("Is entity", () => {
      expect(Internal.Entity.is(imp)).toBeTruthy();
    });
  });

  test("Families", () => {
    const data = Entity.Process(EntityDoc);

    expect(data).toMatchSnapshot();
  });
});
