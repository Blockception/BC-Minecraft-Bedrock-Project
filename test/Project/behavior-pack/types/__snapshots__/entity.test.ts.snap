// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Entity Families 1`] = `
{
  "animations": {
    "defined": [
      "controller",
      "something",
    ],
    "using": [
      "controller.animation.chicken",
      "controller.i.dont.exist",
    ],
  },
  "documentation": "BP Entity: blockception:sheep",
  "events": [
    "self:to",
  ],
  "families": [
    "animal",
    "mob",
  ],
  "groups": [
    "self:group1",
  ],
  "id": "blockception:sheep",
  "location": {
    "position": 98,
    "uri": "C:\\temp.json",
  },
  "molang": {
    "contexts": {
      "using": [],
    },
    "queries": {
      "using": [],
    },
    "temps": {
      "defined": [],
      "using": [],
    },
    "variables": {
      "defined": [],
      "using": [],
    },
  },
  "properties": [],
}
`;
